collector/exporter.go

```diff
@@ -23,7 +25,10 @@ const (
 
 // Metric descriptors.
 var (
-	scrapeDurationDesc = prometheus.NewDesc(
+	db                 *sql.DB
```

Comment:
I guess there is not much benefit if we make it part of the exporter object. Then we will need to pass the exporter object to all the functions which need access to DB and may need changes in existing functions for that.

Reply:
Having the DB as a global brings multiple downsides.

1. It's only ever used by the Exporter and depends on fields of the Exporter. Having it as a global hides the relationship of the Exporter and the DB pool (the Exporter basically "owns" the pool).
2. Currently the API allows you to have multiple, independent Exporter instances that collect from different clusters/pools. By depending on a global variable you break the API since all Exporters share one pool. Even worse this is not visible from the API, since you can still create new Exporters by calling New with different DSNs, but  

You can just pass the db field of the Exporter to the Scrape* functions like this

```
ScrapePerfEventsWaits(e.db, ch)
```

This way you don't need to change much and still avoid the global.

Reply:
Okay. I will make the change. Globals are not good in general.

But I guess, one mysqld_exporter instance can collect from one mysql server only; at least that is what I understood from code. As DATA_SOURCE_NAME env variable is to be defined like user:password@host:port/database. And also the function parseMycnf in mysqld_exporter.go is getting the configs for one DB server only.

Reply:
From the point of the exporter as application you are right, but the nice thing about keeping the API as is, is that the whole collector logic can be used as a library in other tools and even extended. Also in most cases separating application and logic like this leads to cleaner code overall. And who knows, maybe someday the exporter will be changed to collect from multiple sources and then the collector package doesn't need to change.

Reply:
The exporter will never collect from multiple sources.  This violates Prometheus best practices.  Exporters are supposed to be 1:1 with their targets.

Reply:
Yes. I agree.

Reply:
You're right, but avoiding the global state is still a good idea IMO. The multiple sources were just an example albeit not a good one.